src/main/kotlin/grpc/AuthenticationInterceptor.kt

```diff
@@ -0,0 +1,298 @@
+
+        override fun <ReqT : Any?, RespT : Any?> interceptCall(
+            call: ServerCall<ReqT?, RespT?>?,
+            headers: Metadata?,
+            next: ServerCallHandler<ReqT?, RespT?>?,
+        ): ServerCall.Listener<ReqT?>? {
+            val methodName = call?.methodDescriptor?.fullMethodName
+            logger.debug { "Auth Interceptor: Intercepting call to $methodName" }
+
+            if (methodName == null) {
+                call?.close(Status.UNAUTHENTICATED.withDescription("Method name is null"), Metadata())
+                return emptyListener()
+            }
+
+            // Bypass authentication for public methods / health checks
+            if (methodName in PUBLIC_METHODS || methodName.startsWith("grpc.health.v1.Health")) {
+                return next?.startCall(call, headers)
+            }
+
+            // All other methods are protected
+            val cookieHeader = headers?.get(GrpcContext.COOKIE_METADATA_KEY)
+            val cookies = parseCookies(cookieHeader)
+            val accessToken = cookies[GrpcContext.ACCESS_TOKEN_COOKIE_NAME]
+            val refreshToken = cookies[GrpcContext.REFRESH_TOKEN_COOKIE_NAME]
+
+            // Handle the RenewSession call specifically
+            if (methodName == "snowballr.SnowballR/RenewSession") {
+                return handleTokenRefresh(refreshToken, call, headers, next)
+            }
+
+            // For other protected methods, first check access token
+            if (accessToken != null && JWTUtils.isTokenValid(accessToken)) {
+                return attachUserIdAndContinue(accessToken, call, headers, next)
+            }
+
+            // If access token is invalid or missing, try to refresh
+            logger.debug { "Access token invalid or missing for $methodName. Attempting refresh." }
+            return handleTokenRefresh(refreshToken, call, headers, next)
+        }
+
+        /**
+         * Validates the provided access token, extracts the user ID, and attaches it to the gRPC [Context].
+         *
+         * If the token is valid and the user ID is successfully parsed, the context is updated with
+         * [GrpcContext.USER_ID_CONTEXT_KEY], and the call proceeds using the [next] handler.
+         * If token parsing fails (e.g., malformed user ID), the call is closed with [Status.UNAUTHENTICATED].
+         *
+         * @param ReqT The request type.
+         * @param RespT The response type.
+         * @param token The JWT access token to validate and parse.
+         * @param call The current gRPC server call.
+         * @param headers The metadata headers associated with the call.
+         * @param next The next handler in the interceptor chain.
+         * @return A [ServerCall.Listener] to proceed with the call, or a no-op listener if authentication fails.
+         */
+        private fun <ReqT : Any?, RespT : Any?> attachUserIdAndContinue(
+            token: String,
+            call: ServerCall<ReqT?, RespT?>?,
```

Comment:
This is always non-null when we call it.
```suggestion
            call: ServerCall<ReqT?, RespT?>,
```

Reply:
True